File: platform/pkg/observability/metrics/gauge.go
// platform/pkg/observability/metrics/gauge.go
package metrics

import (
	"sync"
)

// Gauge is a metric that can go up and down.
// Used for tracking memory usage, queue size, active connections, etc.
type Gauge interface {
	// Set sets the gauge to the given value
	Set(value float64)

	// Inc increments the gauge by 1
	Inc()

	// Dec decrements the gauge by 1
	Dec()

	// Add adds the given value to the gauge
	Add(delta float64)

	// Sub subtracts the given value from the gauge
	Sub(delta float64)

	// Value returns the current gauge value
	Value() float64

	// Name returns the gauge name
	Name() string

	// Labels returns the gauge labels
	Labels() map[string]string
}

// gauge implements Gauge.
type gauge struct {
	name   string
	help   string
	labels map[string]string
	value  float64
	mu     sync.RWMutex
}

// NewGauge creates a new gauge.
func NewGauge(name, help string, labels map[string]string) Gauge {
	if labels == nil {
		labels = make(map[string]string)
	}
	return &gauge{
		name:   name,
		help:   help,
		labels: labels,
		value:  0,
	}
}

// Set sets the gauge to the given value.
func (g *gauge) Set(value float64) {
	g.mu.Lock()
	defer g.mu.Unlock()
	g.value = value
}

// Inc increments the gauge by 1.
func (g *gauge) Inc() {
	g.Add(1)
}

// Dec decrements the gauge by 1.
func (g *gauge) Dec() {
	g.Sub(1)
}

// Add adds the given value to the gauge.
func (g *gauge) Add(delta float64) {
	g.mu.Lock()
	defer g.mu.Unlock()
	g.value += delta
}

// Sub subtracts the given value from the gauge.
func (g *gauge) Sub(delta float64) {
	g.Add(-delta)
}

// Value returns the current gauge value.
func (g *gauge) Value() float64 {
	g.mu.RLock()
	defer g.mu.RUnlock()
	return g.value
}

// Name returns the gauge name.
func (g *gauge) Name() string {
	return g.name
}

// Labels returns the gauge labels.
func (g *gauge) Labels() map[string]string {
	return g.labels
}

File: platform/pylib/scout_common/observability/metrics/gauge.py
# platform/pylib/scout_common/observability/metrics/gauge.py
"""Gauge metrics for values that can go up and down."""

import threading
from typing import Dict, Optional


class Gauge:
    """Gauge is a metric that can go up and down.
    
    Used for tracking memory usage, queue size, active connections, etc.
    
    Example:
        gauge = Gauge("active_connections", "Number of active connections")
        gauge.inc()        # 1
        gauge.add(5)       # 6
        gauge.dec()        # 5
        gauge.set(10)      # 10
        print(gauge.value) # 10
    """

    def __init__(
        self,
        name: str,
        help: str,
        labels: Optional[Dict[str, str]] = None
    ):
        """Initialize gauge.
        
        Args:
            name: Metric name
            help: Help description
            labels: Label key-value pairs
        """
        self._name = name
        self._help = help
        self._labels = labels or {}
        self._value = 0.0
        self._lock = threading.RLock()

    def set(self, value: float) -> None:
        """Set the gauge to the given value.
        
        Args:
            value: Value to set
        """
        with self._lock:
            self._value = value

    def inc(self) -> None:
        """Increment gauge by 1."""
        self.add(1)

    def dec(self) -> None:
        """Decrement gauge by 1."""
        self.sub(1)

    def add(self, delta: float) -> None:
        """Add the given value to the gauge.
        
        Args:
            delta: Value to add
        """
        with self._lock:
            self._value += delta

    def sub(self, delta: float) -> None:
        """Subtract the given value from the gauge.
        
        Args:
            delta: Value to subtract
        """
        self.add(-delta)

    @property
    def value(self) -> float:
        """Get current gauge value."""
        with self._lock:
            return self._value

    @property
    def name(self) -> str:
        """Get gauge name."""
        return self._name

    @property
    def help(self) -> str:
        """Get gauge help text."""
        return self._help

    @property
    def labels(self) -> Dict[str, str]:
        """Get gauge labels."""
        return self._labels.copy()

    def __repr__(self) -> str:
        """Get debug representation."""
        labels_str = ", ".join(f"{k}={v}" for k, v in self._labels.items())
        return f"Gauge({self._name}{{{labels_str}}}={self._value})"